models/transformer/decoder.hpp

```diff
@@ -0,0 +1,150 @@
+/**
+ * @file models/transformer/decoder.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Definition of the Transformer Decoder layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MODELS_TRANSFORMER_DECODER_HPP
+#define MODELS_TRANSFORMER_DECODER_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/layer/layer_types.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/regularizer/no_regularizer.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * In addition to the two sub-layers in each encoder layer, the decoder inserts
+ * a third sub-layer, which performs multi-head attention over the output of the
+ * encoder stack. Similar to the encoder, we employ residual connections around
+ * each of the sub-layers, followed by layer normalization. We also modify the
+ * self-attention sub-layer in the decoder stack to prevent positions from
+ * attending to subsequent positions. This masking, combined with fact that the
+ * output embeddings are offset by one position, ensures that the predictions
+ * for position i can depend only on the known outputs at positions less than i.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam ActivationFunction The type of the activation function to be used in
+ *         the position-wise feed forward neural network.
+ * @tparam RegularizerType The type of regularizer to be applied to layer
+ *         parameters.
+ */
+template <
+  typename InputDataType = arma::mat,
+  typename OutputDataType = arma::mat,
+  class ActivationFunction = ReLULayer<>,
+  typename RegularizerType = NoRegularizer
+>
+class TransformerDecoder
+{
+ public:
+  typedef typename InputDataType::elem_type ElemType;
+  //! Create the TransformerDecoder object.
+  TransformerDecoder();
+
+  /**
+   * Create the TransformerDecoder object using the specified parameters.
+   *
+   * @param dModel The number of features in the input. Also, same as the
+   *        `embedDim` in `MultiheadAttention` layer.
+   * @param numHeads The number of attention heads.
+   * @param dimFFN The dimentionality of feedforward network.
+   * @param dropout The dropout rate.
+   */
+  TransformerDecoder(const size_t dModel = 512,
```

Comment:
Does d in Model stand for dimension? Maybe we can find a more descriptive name, e.g. `modelDimension`  or `numFeatures`?

Reply:
Yup! Sure. Actually the [paper](https://arxiv.org/abs/1706.03762) used "d<sub>model</sub>" everywhere. As mentioned in the argument description, `dModel` is same as `embedDim` (size of each embedding vector). Should I make it `embedDim` (or `embeddingDim` or `embeddingDimension`) as it says more about what actually it is.